packages/aztec.js/src/proof/proof.js

```diff
@@ -0,0 +1,184 @@
+const { constants, errors } = require('@aztec/dev-utils');
+const BN = require('bn.js');
+const { Enum } = require('enumify');
+const { padLeft } = require('web3-utils');
+
+const bn128 = require('../bn128');
+const Keccak = require('../keccak');
+const ProofUtils = require('./utils');
+const types = require('./types');
+
+const { AztecError } = errors;
+
+class ProofType extends Enum {}
+ProofType.initEnum(types);
+
+/**
+ * @class
+ * @class Class to create, store and pass around AZTEC proofs and related transcripts, outputs and hashes.
+ */
+class Proof {
+    /**
+     * @param {string} type one of the ProofType enum values
+     * @param {Note[]} inputNotes input AZTEC notes
+     * @param {Note[]} outputNotes output AZTEC notes
+     * @param {string} sender Ethereum address of transaction sender
+     * @param {string} publicValue public commitment being added to proof
+     * @param {string} publicOwner holder of a public token being converted
+     * @param {Object} metadata arbitrarily-sized object
+     */
+    constructor(type, inputNotes, outputNotes, sender, publicValue, publicOwner, metadata = []) {
+        this.type = type;
+        this.inputNotes = inputNotes;
+        this.m = inputNotes.length;
+        this.outputNotes = outputNotes;
+        this.notes = [...inputNotes, ...outputNotes];
+        this.sender = sender;
+        this.publicOwner = publicOwner;
+        this.metadata = metadata;
+        if (BN.isBN(publicValue)) {
+            this.publicValue = publicValue;
+        } else if (publicValue < 0) {
+            this.publicValue = bn128.curve.n.sub(new BN(-publicValue));
```

Comment:
This file uses `bn128.curve.n` where we should probably use `GROUP_MODULUS`

Reply:
Worth adding, arithmetic over BN numbers that use `GROUP_MODULUS` as their 'reduction context' is not compatible with BN numbers that use `bn128.curve.n`, despite them having the same value (BN.js checks that the reduction context is 'the same', but it does this via an equality check, so what's actually being checked is the memory location of the reduction context).  

i.e. we should pick either `GROUP_MODULUS` or `bn128.curve.n` and use that value consistently across aztec.js